scripts/iso_mirna_quantification.py

```diff
@@ -0,0 +1,337 @@
+            "the output table. Default: %(default)s."
+        ),
+        action='store_true',
+        default=False
+    )
+
+    return parser
+
+
+def get_out_dirs(mir_list: list[str], library: str, outdir: Path) -> Optional[Path]:
+    """Get output directories.
+    
+    Create directories for each miRNA type specified in a list and return a
+    the new directories. If a miRNA type is not found, its corresponding value
+    in the returned list is None.
+
+    Args:
+        mir_list:
+            list with the desired miRNA types tables
+        library:
+            library where the species to be classified belong to
+        outdir:
+            path to the output directory
+    Returns:
+        the new directories and/or None
+    """
+    mir_out = None
+    isomir_out = None
+    iso_mirna_out = None
+
+    if 'mirna' in mir_list:
+        mir_out = outdir/f'mirna_counts_{library}'
+    if 'isomir' in mir_list:
+        isomir_out = outdir/f'isomir_counts_{library}'
+    if 'iso_mirna' in mir_list:
+        iso_mirna_out = outdir/f'iso_mirna_counts_{library}'
+
+    return mir_out, isomir_out, iso_mirna_out
+
+
+def get_contribution(aln: pysam.AlignedSegment, collapsed: bool = False, nh: bool = False) -> float:
+    """Get contribution of an alignment to the overall count.
+
+    Calculate the contribution of the alignment as the number of reads divided
+    by NH. The arguments `collapsed` and `nh` determine if these values are
+    taken from the alignment name. If both arguments are set to `True`, the
+    contribution of each alignment is computed as the number of reads/NH and
+    the values are extracted from the name. If only `collapsed` is `True`, the
+    contribution is the number of reads/NH tag. If only `nh` is `True`, the
+    contribution is 1/NH. Otherwise, the contribution is 1/NH tag. If `nh` is
+    `False` and the alignment misses the NH tag, the value is set to 1.
+
+    Args:
+        aln:
+            alignment to calculate the contribution from
+        collapsed:
+            indicates if the alignment is collapsed or not
+        nh:
+            indicates if the NH vale is found on the alignment name
+    Returns:
+        the contribution of the alignment to the overall total
+    """
+    query_id = aln.query_name
+
+    if collapsed and nh:
+        num_reads = int(query_id.split('-')[1].split('_')[0])
+        nh_value = int(query_id.split('-')[1].split('_')[1])
+
+    elif not collapsed and nh:
+        num_reads = 1
+        nh_value = int(query_id.split('_')[1])
+
+    elif collapsed and not nh:
+        num_reads = int(query_id.split('-')[1])
+        try:
+            nh_value = aln.get_tag("NH")
+        except KeyError:
+            nh_value = 1
+
+    else:
+        num_reads = 1
+        try:
+            nh_value = aln.get_tag("NH")
+        except KeyError:
+            nh_value = 1
+
+    return num_reads/nh_value
```

Comment:
This algorithm will be super slow for cases where neither NH nor the number of collapsed reads are used. It will actually be the slowest of all possible modes, even though it _should_ be the fastest (because, if the options are _not_ indicated, we should never need to check that for any read).

Best is if you can find a way to avoid all these conditionals completely in the loop. You could generate a function that always returns you the values for NH and the number of counts either given a regex or a default, then map that function as an input to a general counting function that simply does `num_reads/nh_value` to your alignments.

Reply:
Basically, there are roughly two approaches you can follow:

1. You define 4 different counting/contribution functions, one each for the combined states of `collapsed` and `nh` (which either can be `True` or `False`). Then, _before_ you loop over the individual alignments, you determine which of the four functions is the function you need to use here. Then when looping over the alignments, you apply that function to your read.
2. Similar approach but you only define one very simple counting/contribution function that takes values for number of collapsed reads and the number of hits inputs and returns their quotient, i.e., $n_{reads}/n_{hits}$. And then you have to populate the input params with an appropriate regex. But as you would need to fall back to defaults if NH or number of reads is not part of the query name, you'd probably end up having to write small functions for each case here as well, so the benefit over 1 may be minimal. Besides, in three of the four cases defaults (i.e., 1) are used for either or both of the two params, so always calculating the quotient is not necessary or at least will carry the overhead of defining a variable where none is needed.